lib/pages/auth/login_page.dart

```diff
@@ -0,0 +1,200 @@
+  bool _isLoading = false;
+  int _loginAttempts = 0;
+  DateTime? _lastFailedLogin;
+
+  String _hashPassword(String password) {
+    final bytes = utf8.encode(password);
+    final hash = sha256.convert(bytes);
+    return hash.toString();
+  }
+
+  Future<void> _loginWithPhone(String phoneNumber, String password) async {
+    if (_lastFailedLogin != null) {
+      final difference = DateTime.now().difference(_lastFailedLogin!);
+      if (difference.inMinutes < 1) {
+        if (mounted) {
+          ScaffoldMessenger.of(context).showSnackBar(
+            SnackBar(
+              content: Text(
+                'Too many attempts. Please try again in ${60 - difference.inSeconds} seconds',
+              ),
+              backgroundColor: Colors.red,
+            ),
+          );
+        }
+        return;
+      } else {
+        _loginAttempts = 0;
+        _lastFailedLogin = null;
+      }
+    }
+
+    setState(() {
+      _isLoading = true;
+    });
+
+    try {
+      String cleanPhoneNumber = '+${phoneNumber.trim()}';
+      if (!cleanPhoneNumber.startsWith('+')) {
+        cleanPhoneNumber = '+$cleanPhoneNumber';
+      }
+
+      final querySnapshot =
+          await FirebaseFirestore.instance
+              .collection('users')
+              .where('phone', isEqualTo: cleanPhoneNumber)
+              .get();
+
+      if (querySnapshot.docs.isEmpty) {
+        _handleFailedLogin();
+        throw Exception("Phone number not found");
+      }
+
+      final userData = querySnapshot.docs.first.data();
+      final storedPassword = userData['password'];
+      final hashedInputPassword = _hashPassword(password);
+
+      if (storedPassword != hashedInputPassword) {
+        _handleFailedLogin();
+        throw Exception("Phone number or password is incorrect");
+      }
+
+      final userProvider = Provider.of<UserProvider>(context, listen: false);
+
+      _loginAttempts = 0;
+      _lastFailedLogin = null;
+
+      if (mounted) {
+        await userProvider.login(phoneNumber, password);
+        final prefs = await SharedPreferences.getInstance();
+        await prefs.setString('logged_in_phone', phoneNumber);
+
+        if (userProvider.currentUser!.type == UserType.rider) {
+          Navigator.pushReplacement(
+            context,
+            MaterialPageRoute(builder: (context) => const RiderScreen()),
+          );
+        }else{
+          Navigator.pushReplacement(
+            context,
+            MaterialPageRoute(builder: (context) => const DriverMap()),
+          );
+        }
+      }
+    } catch (e) {
+      if (mounted) {
+        ScaffoldMessenger.of(context).showSnackBar(
+          SnackBar(content: Text(e.toString()), backgroundColor: Colors.red),
+        );
+      }
+    } finally {
+      if (mounted) {
+        setState(() {
+          _isLoading = false;
+        });
+      }
+    }
+  }
+
+  void _handleFailedLogin() {
+    _loginAttempts++;
+    if (_loginAttempts >= 5) {
```

Comment:
maxAttempts = 5 to avoid magic number and it will be much better if it is configurable or stored in firebase